### PR TITLE
clippy lints

### PR DESCRIPTION
**Pull request by owenthewizard**
Many many lints. Notably:
- Remove unneeded `Result` wrapping some function return types.
- `if ... { panic! }` to `assert!`
- Replaced `if let` with `map_or` and similar.
- [Make `fn` `const` if possible.](https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn)
- [`ptr::addr_of!` instead of `&T as *const T` ](https://rust-lang.github.io/rust-clippy/master/index.html#borrow_as_ptr)
- [`.cast() instead of `*const T as *const U`](https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr)
- [`.cast_mut()` instead of `*const T as *mut T`](https://rust-lang.github.io/rust-clippy/master/index.html#ptr_cast_constness)
- [`U::from(T)` instead of `T as U`](https://rust-lang.github.io/rust-clippy/master/index.html#cast_lossless)
- `"{}". foo` -> `"{foo}"`
- [Use `Self`](https://rust-lang.github.io/rust-clippy/master/index.html#use_self)
- ...

As many of these are subjective, feel free to modify/request modifications.

### CODE REVIEW

**Comment by ivmarkov** [comment]
> Many many lints. Notably:
> 
> * Remove unneeded `Result` wrapping some function return types.

I would accept this only for non-public API. Can you check and make sure you've done this change for private and crate-private methods only? If Clippy complains in public stuff, we should suppress its warnings there.

> * `if ... { panic! }` to `assert!`
> * Replaced `if let` with `map_or` and similar.
> * [Make `fn` `const` if possible.](https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn)
> * [`ptr::addr_of!` instead of `&T as *const T` ](https://rust-lang.github.io/rust-clippy/master/index.html#borrow_as_ptr)
> * [`.cast() instead of `*const T as *const U`](https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr)
> * [`.cast_mut()` instead of `*const T as *mut T`](https://rust-lang.github.io/rust-clippy/master/index.html#ptr_cast_constness)
> * [`U::from(T)` instead of `T as U`](https://rust-lang.github.io/rust-clippy/master/index.html#cast_lossless)
> * `"{}". foo` -> `"{foo}"`
> * [Use `Self`](https://rust-lang.github.io/rust-clippy/master/index.html#use_self)
> * ...
> 
> As many of these are subjective, feel free to modify/request modifications.

The rest is fine. I don't have strong feelings for doing it one way or the other. These are stylistic changes and as you say, sometimes subjective.
As long as Clippy stays silent, I'm happy.

**Comment by ivmarkov** [comment]
Actually one more thing: None of your improvements seem to address the actual clippy complaints which are currently failing the CI? All of the failures in the CI seem to be related to Clippy demanding that we use explicit turbofish typecasts when using `core::mem::transmute`. I think these should be addressed with priority.

**Comment by ivmarkov** [comment]
And... your PR has come flicts. You need to rebase your PR on top of master. As in `git checkout <your-branch>`. Then `git pull`. Then `git rebase master --interactive`. ... and resolve all conflicts and upon resolving `git rebase --continue` until no conflicts are left and all your commits are on top of master.

**Comment by ivmarkov** [comment]
... where `upstream/master` is the upstream master branch of esp-idf-svc. You can create the upstream remote repo with `git remote add https://github.com/esp-rs/esp-idf-svc`.

**Comment by ivmarkov** [comment]
Ok conflicts seem to be fixed now, but clippy is still not happy. See the failing CI. I think given that this PR addresses Clippy lints, the CI should -as a minimum - be passing.

File: src/eth.rs [hunk 1]
```diff
@@ -259,10 +259,10 @@ impl<'d> EthDriver<'d, RmiiEth> {
         chipset: RmiiEthChipset,
         reset: Option<i32>,
         phy_addr: Option<u32>,
-    ) -> Result<*mut esp_eth_phy_t, EspError> {
```
Comment (owenthewizard) [inline src/eth.rs:262]: Removed uneeded `Result` (private API).

File: src/eth.rs [hunk 2]
```diff
@@ -717,14 +715,14 @@ impl<'d, T> EthDriver<'d, T> {
     ///
     /// This "local borrowing" will only be possible to express in a safe way once/if `!Leak` types
     /// are introduced to Rust (i.e. the impossibility to "forget" a type and thus not call its destructor).
-    pub unsafe fn set_nonstatic_rx_callback<F>(&mut self, callback: F) -> Result<(), EspError>
```
Comment (owenthewizard) [inline src/eth.rs:720]: Same as above (public API).

File: src/eth.rs [hunk 3]
```diff
@@ -717,14 +715,14 @@ impl<'d, T> EthDriver<'d, T> {
...
+    pub unsafe fn set_nonstatic_rx_callback<F>(&mut self, callback: F)
     where
         F: FnMut(EthFrame) + Send + 'd,
     {
-        self.internal_set_rx_callback(callback)
+        self.internal_set_rx_callback(callback);
     }
 
-    fn internal_set_rx_callback<F>(&mut self, callback: F) -> Result<(), EspError>
```
Comment (owenthewizard) [inline src/eth.rs:727]: Private API, but needs to be changed to match public.

File: src/wifi.rs [hunk 4]
```diff
@@ -394,7 +394,7 @@ pub trait NonBlocking {
 
     fn stop_wps(&mut self) -> Result<WpsStatus, EspError>;
 
-    fn is_wps_finished(&self) -> Result<bool, EspError>;
```
Comment (owenthewizard) [inline src/wifi.rs:397]: Remove unneeded `Result` (private API).

File: src/wifi.rs [hunk 5]
```diff
@@ -436,7 +436,7 @@ where
         (**self).stop_wps()
     }
 
-    fn is_wps_finished(&self) -> Result<bool, EspError> {
```
Comment (owenthewizard) [inline src/wifi.rs:439]: As above.

File: src/wifi.rs [hunk 6]
```diff
@@ -1175,12 +1180,12 @@ impl<'d> WifiDriver<'d> {
         }
     }
 
-    fn is_wps_finished(&self) -> Result<bool, EspError> {
```
Comment (owenthewizard) [inline src/wifi.rs:1178]: As above.

File: src/wifi.rs [hunk 7]
```diff
@@ -1711,7 +1716,7 @@ impl<'d> EspWifi<'d> {
         self.driver_mut().stop_wps()
     }
 
-    pub fn is_wps_finished(&self) -> Result<bool, EspError> {
```
Comment (owenthewizard) [inline src/wifi.rs:1714]: As above.

File: src/wifi.rs [hunk 8]
```diff
@@ -1805,7 +1806,7 @@ impl<'d> NonBlocking for EspWifi<'d> {
         EspWifi::stop_wps(self)
     }
 
-    fn is_wps_finished(&self) -> Result<bool, EspError> {
```
Comment (owenthewizard) [inline src/wifi.rs:1808]: As above.

File: src/eth.rs [hunk 9]
```diff
@@ -687,11 +685,11 @@ impl<'d, T> EthDriver<'d, T> {
         Ok(())
     }
 
-    pub fn set_rx_callback<F>(&mut self, callback: F) -> Result<(), EspError>
```
[the remark this answers is not in the document]
Comment (ivmarkov) [reply src/eth.rs:690]: Yes please.

File: src/eth.rs (hunk 2 above)
Comment (ivmarkov) [reply src/eth.rs:720]: Revert.

**Review by ivmarkov — CHANGES_REQUESTED** [review]
As mentioned, please revert the signatures of all public methods. The rest I'll accept and merge, though not sure what it gives us, as I think Clippy was only triggering on the `transmute` calls?

Last but not least and very important: I really do hope you've implemented these changes automatically with `cargo clippy --fix` or similar, as so.e of these are not so trivial, and if these were done manually, they could I trounce bugs, as I don't th8nk you've tested all changed code paths. Can you confirm that you've done these changes in an automatic fashion, and how exactly?

**Comment by owenthewizard** [comment]
> Can you confirm that you've done these changes in an automatic fashion, and how exactly?

Unfortunately not as `clippy --fix` only works with a small subset of lints. Additionally it can't reason about `as *const _ as *mut _` -> `.cast().cast_mut()` without type hints. If you don't want to merge the changes due to this I completely understand. If you like I could submit multiple smaller PRs so the changes can be more easily reviewed.

Would you consider removing the unnecessary `Result` on the `EthDriver` callbacks in the next major/API update?

**Comment by ivmarkov** [comment]
> > Can you confirm that you've done these changes in an automatic fashion, and how exactly?
> 
> Unfortunately not as `clippy --fix` only works with a small subset of lints. Additionally it can't reason about `as *const _ as *mut _` -> `.cast().cast_mut()` without type hints. If you don't want to merge the changes due to this I completely understand. If you like I could submit multiple smaller PRs so the changes can be more easily reviewed.

Splitting into multiple smaller PRs would help then. 

> Would you consider removing the unnecessary `Result` on the `EthDriver` callbacks in the next major/API update?

Not sure. Why do you need that so much?

**Comment by owenthewizard** [comment]
> Not sure. Why do you need that so much?

I don't need it per se, I just think it's bloat since it always returns `Ok(())`.

**Comment by ivmarkov** [comment]
> > Not sure. Why do you need that so much?
> 
> I don't need it per se, I just think it's bloat since it always returns `Ok(())`.

I think this is insignificant in the grant scheme of things. And it is not worth it to break backwards compatibility - even for a new API - because of this. Not to mention that in future we might need a result-returning API after all.

**Comment by owenthewizard** [comment]
Closing this, will split into smaller PRs.

**Pull request closed without merging**